### ISSUE TITLE
Wifi Not working

### ISSUE THREAD

**Issue by chabad360** [billkenney/update_max3_plus3#6]
After installing this image on the X-Max 3, the wifi module stops working (it seems that it no longer want to modeswitch, or rather the driver doesn't work correctly). If I allow armbian to update the firmware, the wifi module seems to modeswitch, but NetworkManager still doesn't see it. However, after the firmware update, ethernet also stops working, rendering the device inaccessible over the network, forcing me to connect over serial.

**Comment by chabad360** [comment]
I had success following [this guide](https://forum.beagleboard.org/t/success-with-brostrend-usb-wifi-dongle-somewhere-to-document-the-process/37007/3). It seems that some other driver with the same name is already installed (but it has an `invalid exec format` error when loading), so just run `dkms install aic8800/1.0.5 --force` to install the correct driver and then you should be good.

If you can figure out a driver fix in your image that would be even better.

**Comment by billkenney** [comment]
> I had success following [this guide](https://forum.beagleboard.org/t/success-with-brostrend-usb-wifi-dongle-somewhere-to-document-the-process/37007/3). It seems that some other driver with the same name is already installed (but it has an `invalid exec format` error when loading), so just run `dkms install aic8800/1.0.5 --force` to install the correct driver and then you should be good.
> 
> If you can figure out a driver fix in your image that would be even better.

did you install the image or follow the manual installation guide? I don't use wifi, but if you're having issues, you could try installing the drivers again (see below). 

I do recall some weird issue about modeswitch from usb mode or something, I don't recall what I followed, but I do remember that my wifi dongle already had right setting, so the modeswitch didn't work. I had assumed others would also be in the right mode... if that's not the case, maybe you could modeswitch prior to installing the image and wifi would continue working

also, did you have to install the wpa-supplicant file in the guide you mentioned? if so, I could probably add that and the firmware to my image with instructions to edit it for your country, network, and password. 

step 17 in the manual install guide: to get your wifi working again, run the following commands: `sudo ln -s /usr/src/linux-headers-6.7.5-edge-rockchip64/arch/arm64 /usr/src/linux-headers-6.7.5-edge-rockchip64/arch/aarch64 ; sudo apt install make gcc build-essential git network-manager ; cd /home/mks ; git clone https://github.com/lwfinger/rtl8xxxu ; cd /home/mks/rtl8xxxu ; make clean modules ; sudo make install ; sudo make install_fw`. Restart the printer, then: `cd /home/mks/rtl8xxxu ; sudo modprobe rtl8xxxu_git`. you can run `sudo nmtui` to use the network-manager service to create or manage network connections. if you get any errors, run `cd /home/mks/rtl8xxxu ; make clean modules ; sudo make install ; sudo make install_fw`, reboot, and run `cd /home/mks/rtl8xxxu ; sudo modprobe rtl8xxxu_git` again

maybe also run `sudo lsusb` and make sure your wifi dongle is the same model? I seriously doubt qidi would swap it out, but I guess it's possible
<img width="896" alt="Screen Shot 2024-06-17 at 9 05 27 PM" src="https://github.com/billkenney/update_max3_plus3/assets/30010560/58744c86-cb0c-40c4-9586-52e5c255c0f8">

**Comment by chabad360** [comment]
I did the image install. 

The wpa supplicant thing didn't help, because the wifi module was in storage mode. Your instructions for building drivers didn't help, because I don't have a realtek module, I have an AIC one. I had to rebuild the AIC8800 driver as the bundled one had an `exec format error` when running `modprobe aic8800_fdrv`.

**Comment by billkenney** [comment]
> I did the image install.
> 
> The wpa supplicant thing didn't help, because the wifi module was in storage mode. Your instructions for building drivers didn't help, because I don't have a realtek module, I have an AIC one. I had to rebuild the AIC8800 driver as the bundled one had an `exec format error` when running `modprobe aic8800_fdrv`.

what does your wifi module show when you run `sudo lsusb`?

I'm surprised they switched to a different wifi dongle, although the Realtek one is pretty bad. 

I'm testing @CChen616's image right now, and trying to integrate his image with my image. The printer.cfg setup is cleaner, and it has some cool macros like klippain-shaketune and spoolman, but I've also encountered some problems. Like the network tab on the screen does nothing, in fluidd Layer is either the same for every print (and wrong) or it has nothing, and I had the screen completely stopped working at one point. 

If the drivers for the AIC wifi are not in this image, I'll add them and release a new image. Once I figure out the best system setup. 

Thanks for the feedback, and I'm glad you were able to figure out what the problem was

**Comment by chabad360** [comment]
> what does your wifi module show when you run sudo lsusb?

`Bus 002 Device 003: ID 2604:0013 Tenda AIC8800DC`

> I'm surprised they switched to a different wifi dongle, although the Realtek one is pretty bad.

This one is not any better.

> Like the network tab on the screen does nothing,

It doesn't work for me on your image either.

I should note regarding the screen, someone on a discord is working on custom screen firmware (and seems to be pretty close to something working), so IDK how worth it, it is to focus on getting the screen working.

**Comment by billkenney** [comment]
> > Like the network tab on the screen does nothing,
> 
> 
> 
> It doesn't work for me on your image either.

In my image I was always able to go into the network menu, but when you clicked the WiFi icon it didn't work. In this one it doesn't even go into the network menu. But you're right, it's broken regardless of how you update. Although I will say that gcode images work sometimes, albeit very sporadically (I have no idea why they do or do not work).

> I should note regarding the screen, someone on a discord is working on custom screen firmware (and seems to be pretty close to something working), so IDK how worth it, it is to focus on getting the screen working.

The screen is really the only thing preventing people from updating. It would be awesome if someone figured out a way to replace xindi. From what I understand, the program/source code used to compile xindi is all in Chinese, so some people have gotten access to it but they can't really use it because they don't read Chinese. 

FYI @CHhen616 said he didn't include the AIC firmware in his image either, but the drivers were included on an update at some point. I had no idea there were other WiFi modules on these printers, thanks for letting me know! I'll add that to my repo, and if I release another image, I'll install that firmware so both are available.

**Comment by chabad360** [comment]
I may have figured out how to get the wifi screen working. It seems your image doesn't name the wifi iface as `wlan0` (thanks Persistent network interfaces), just adding `extraargs=net.ifacenames=0` to `/boot/armbianEnv.txt` seems to fix that. 

But in regards to making a connection, it's a bit more complicated, in the ui you can put in a password, but xindi seems to get stuck trying to connect. If I manually go into `wpa_cli` and add the connection and then restart xindi, then it works. However, NetworkManager seems to have a conflict with the `makerbase-wlan0` service. This can be fixed by adding `-u` to the end of the command for that service (`-u` allows dbus communication, which NetworkManager needs). That still doesn't make the connection work tho. ~Not sure how that worked on the original image (I may need to swap emmc's)...~ It seems like it may have used dhclient to handle that. Which I'm able to supplement with dhcpcd and enabling `dhcpcd@wlan0.service`, which seems to solve that problem, but it doesn't add dns, which makes it not to usefull. Either way, that's enough for now, I need to take care of other things.

**Comment by billkenney** [comment]
> adding `extraargs=net.ifacenames=0` to `/boot/armbianEnv.txt` seems to fix that. 

This didn't work for me. What did work was: `sudo printf 'ACTION=="add", SUBSYSTEM=="net", KERNEL=="wlan0", NAME="wlan0"' > /etc/udev/rules.d/70-custom-names.rules ; udevadm control --reload-rules ; sudo reboot`

> But in regards to making a connection, it's a bit more complicated, in the ui you can put in a password, but xindi seems to get stuck trying to connect. If I manually go into `wpa_cli` and add the connection and then restart xindi, then it works. However, NetworkManager seems to have a conflict with the `makerbase-wlan0` service. This can be fixed by adding `-u` to the end of the command for that service (`-u` allows dbus communication, which NetworkManager needs). That still doesn't make the connection work tho. ~Not sure how that worked on the original image (I may need to swap emmc's)...~ It seems like it may have used dhclient to handle that. Which I'm able to supplement with dhcpcd and enabling `dhcpcd@wlan0.service`, which seems to solve that problem, but it doesn't add dns, which makes it not to usefull. Either way, that's enough for now, I need to take care of other things.

I did all of this and the network menu on the screen looks like it's working, it sees the networks and lets you put in a password, but it freezes at connecting. 


I did notice that my image has what looks to be aic drivers in /root/xinwangka/aic8800DC. Seems they were installed by the firmware update, but based on the fact that you opened this issue, it didn't work. 
![image](https://github.com/billkenney/update_max3_plus3/assets/30010560/d16ed6f7-984b-446e-a96e-3c1173c94343)

I think maybe the drivers didn't work because they are in the wrong place. My manual install script requires you to create a link from /usr/lib/modules/5.16.20-rockchip64 --> /usr/lib/modules/6.7.5-edge-rockchip64. The firmware update apparently deleted that symlink, and installed the aic drivers in there.

**Comment by chabad360** [comment]
> This didn't work for me. 

Oh, right. You also need to compile the `boot.scr` again.
```sh
sudo mkimage -C none -A arm -T script -d /boot/boot.cmd /boot/boot.scr
```

**Comment by chabad360** [comment]
The only thing that wasn't working was dns, and the pretty simple solution there was to replace `/etc/resolv.conf` with a link to the systemd-resolvd one.

**Comment by billkenney** [comment]
> The only thing that wasn't working was dns, and the pretty simple solution there was to replace `/etc/resolv.conf` with a link to the systemd-resolvd one.

So you're saying delete /etc/resolv.conf and sudo ln -s /etc/systemd/resolved.conf /etc/resolv.conf? I did notice that the resolv.conf file has name servers google I think, and it was generated by network manager. Do you not need name servers? The only thing not commented out in resolved.conf is:
[Resolve]
DNS=1.0.0.1

I wonder is a better solution would be to use a sed command and append those lines to the end of the file?

**Comment by chabad360** [comment]
No, the resolv.conf gets automatically generated, so we dont want to tamper with it, and since NetworkManager isn't managing the wifi interface, it doesn't add nameservers for it. Systemd-resolved on the other hand, doesn't care, and figures it out (probably with thanks to dhcpcd). By swapping the resolv.conf, NM will automatically switch to using systemd-resolved for DNS (so ethernet will work properly too), and all problems are(/should be) solved.

**Comment by billkenney** [comment]
> No, the resolv.conf gets automatically generated, so we dont want to tamper with it, and since NetworkManager isn't managing the wifi interface, it doesn't add nameservers for it. Systemd-resolved on the other hand, doesn't care, and figures it out (probably with thanks to dhcpcd). By swapping the resolv.conf, NM will automatically switch to using systemd-resolved for DNS (so ethernet will work properly too), and all problems are(/should be) solved.

Can you provide me with a terminal command that will fix the issue? If so, I'll test it and add it to the guide. Something I would imagine like `sudo mv /etc/resolv.conf /etc/resolv.conf.bak ; sudo ln -s /etc/systemd/resolved.conf /etc/resolv.conf`

**Comment by chabad360** [comment]
`sudo mv /etc/resolv.conf /etc/resolv.conf.bak ; sudo ln -s /run/systemd/resolve/stub-resolv.conf /etc/resolv.conf` and restarting network manager should do the trick. [You'll still need `dhcpcd` to handle dhcp, and wpa_supplicant configured correctly](https://github.com/billkenney/update_max3_plus3/issues/6#issuecomment-2176847749). Tbh, no reason why none of this can't be preconfigured in the image (although there definitely should be some sort of instructions on how to switch back to NM).

**Comment by billkenney** [comment]
> `sudo mv /etc/resolv.conf /etc/resolv.conf.bak ; sudo ln -s /run/systemd/resolve/stub-resolv.conf /etc/resolv.conf` and restarting network manager should do the trick. [You'll still need `dhcpcd` to handle dhcp, and wpa_supplicant configured correctly](https://github.com/billkenney/update_max3_plus3/issues/6#issuecomment-2176847749). Tbh, no reason why none of this can't be preconfigured in the image (although there definitely should be some sort of instructions on how to switch back to NM).

Does this fix the wifi on the screen? If so, then it's definitely worth integrating into a new image. If not, is there a benefit to using wpa_supplicant over network manager? For instance, if the screen wifi doesn't work, you still need to have ethernet to ssh into your printer. My wpa_supplicant file is completely blank, but I can still connect to wifi with network manager. 

Edit: I ran the command you posted. I am able to see the wifi networks on the screen, select a network, and type the password in, but it freezes when I press the check mark and writes nothing to wpa_supplicant-wlan0.conf. I deleted the wpa_supplicant file and recreated it with the below content, then restarted the printer. I had to delete /var/run/wpa_supplicant/wlan0 and restart makerbase-wlan0.service, and the wifi connects on reboot. But when I try to choose a different network the screen freezes up again and makes no changes to the wpa_supplicant file. 

(I've tried ctrl_interface_group=netdev and ctrl_interface_group=0)

ctrl_interface=/var/run/wpa_supplicant
ctrl_interface_group=netdev
update_config=1

network={
	ssid="XXXXXXX"
	psk="XXXXXXX"
}

What's odd, is that when I restart the makerbase-client service, it says that it is connected to the other network, although that isn't reflected in the wpa_supplicant. It also seems I have to delete /var/run/wpa_supplicant/wlan0 every time I restart the printer. 

Edit2: I was able to get the makerbase-wlan0 service working by adding this: 
ExecStartPre=-/root/delwlan0.sh

Contents of the /root/delwlan0.sh script: 
#!/bin/bash
rm /var/run/wpa_supplicant/wlan0

After reboot the screen shows that I am still connected to the original network, it no longer shows me connected to the network I tried to change to. For some reason these changes are not being saved. 

Edit3: Okay so I rebooted again, tried to change the connection on the screen, got a "Connection successful" message, "saving wifi settings", and it changed wpa_supplicant. I'm not sure what exactly changed, but it seems to be working now. I'll create another image when I have the time. 

Thanks for opening the issue and figuring this out.

**Comment by chabad360** [comment]
> #!/bin/bash
> rm /var/run/wpa_supplicant/wlan0

OMG, this is actually the worst solution to this problem.

Doing it this way breaks things more, I'm pretty sure they (makerbase/qidi) could've just used the same socket as `wpa_supplicant.service` (and slightly adjusted the config) or used `wpa_supplicant@wlan0.service` and slightly modified that service (with `-u`) and it would've worked fine (heck, they could've just relied on NetworkManager in the first place), but they had to do things special. Seriously, the amount of stupid hacks I ran into surrounding wireless networking is starting to drive me mad.

#### What you should do
In `makerbase-wlan0.service`, you can just change
```diff
[Unit]
Description=Makerbase mkspi WPA supplicant daemon
BindsTo=sys-subsystem-net-devices-wlan0.device
- Before=network.target
+ Before=dhcpcd@wlan0.service
- Wants=network.target
+ After=dbus.service

[Service]
Type=simple
- ExecStart=/sbin/wpa_supplicant -c/etc/wpa_supplicant/wpa_supplicant-wlan0.conf -Dnl80211,wext -iwlan0
+ ExecStart=/sbin/wpa_supplicant -c/etc/wpa_supplicant/wpa_supplicant-wlan0.conf -Dnl80211,wext -iwlan0 -u

[Install]
+ WantedBy=network.target
- Alias=multi-user.target.wants/makerbase-wlan0.service
```
then run
```bash
sudo apt install dhcpcd -y
sudo systemctl disable dhcpcd.service
sudo systemctl enable dhcpcd@wlan0.service
sudo systemctl disable wpa_supplicant.service
sudo systemctl enable makerbase-wlan0.service
```

and then it should pretty much work fine without issues, and as bonus, NetworkManager will still be able to use it (I think, but I haven't tried, it might conflict with dhcpcd). 

Basically, `-u` tells wpa_supplicant to expose a DBus interface (that NM relies upon), if NM can't find that socket, it will start wpa_supplicant.service to provide that interface, creating a conflict. Having `wpa_supplicant.service` also be enabled, invites that conflict. My solution navigates having conflicting networking systems running (always a bad idea...).

**Comment by billkenney** [comment]
I tried this and it works. This should work: `sudo mv /etc/resolv.conf /etc/resolv.conf.bak ; sudo ln -s /run/systemd/resolve/stub-resolv.conf /etc/resolv.conf ; sudo printf 'ctrl_interface=/var/run/wpa_supplicant\nctrl_interface_group=0\nupdate_config=1\n\nnetwork={\n  ssid="network"\n  psk="password"\n  mesh_fwding=1\n}\n' > /tmp/wpa ; sudo mv /tmp/wpa /etc/wpa_supplicant/wpa_supplicant-wlan0.conf ; sudo chown root:root /root/.bash_history /root/.zsh_history /etc/wpa_supplicant/wpa_supplicant-wlan0.conf ; sudo sed -Ei 's/^After.*$/Before=dhcpcd@wlan0.service\nAfter=dbus.service/;s/^ExecStart=.*$/ExecStart=\/sbin\/wpa_supplicant -c\/etc\/wpa_supplicant\/wpa_supplicant-wlan0.conf -Dnl80211,wext -iwlan0 -u/;s/^Alias=.*$/WantedBy=network.target/' /lib/systemd/system/makerbase-wlan0.service ; sudo sed -Ei 's/^(Description.*$)/\1\nAfter=mariadb.service/' /etc/systemd/system/Spoolman.service ; sudo systemctl daemon-reload ; sudo apt install dhcpcd -y ; sudo systemctl disable dhcpcd.service ; sudo systemctl enable dhcpcd@wlan0.service ; sudo systemctl disable wpa_supplicant.service ; sudo systemctl enable makerbase-wlan0.service ; sudo reboot`

**Comment by chabad360** [comment]
Why do you have the commands for chown .bash_history and editing the spoolman config in that line? Otherwise, LGTM.

**Comment by billkenney** [comment]
> Why do you have the commands for chown .bash_history and editing the spoolman config in that line? Otherwise, LGTM.

I noticed that the history files were owned by mks on my machine (and therefore in my image). I deleted them and created blank copies when I made my last image to delete my history. I doubt it would cause problems, but they should be owned by root. 

I also added spoolman to my image, and I configured it to use Mariadb because psql wasn't working for me. In the syslog I saw errors on reboot because spoolman couldn't connect to mariadb, because it wasn't running yet. 

I figured I'd incorporate those into this step, because I don't know when I'm going to create an updated image.
